File: backend/src/main/java/com/urjc/plushotel/services/ReservationService.java
package com.urjc.plushotel.services;

import com.urjc.plushotel.dtos.request.ReservationRequest;
import com.urjc.plushotel.dtos.response.ReservationDTO;
import com.urjc.plushotel.dtos.response.ReservedDatesDTO;
import com.urjc.plushotel.entities.*;
import com.urjc.plushotel.exceptions.InvalidReservationRangeException;
import com.urjc.plushotel.exceptions.ReservationNotFoundException;
import com.urjc.plushotel.repositories.ReservationRepository;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;

import java.security.SecureRandom;
import java.util.List;

@Service
public class ReservationService {

    private final ReservationRepository reservationRepository;
    private final RoomService roomService;
    private final CustomUserDetailsService userDetailsService;

    private static final SecureRandom RANDOM = new SecureRandom();

    public ReservationService(ReservationRepository reservationRepository, RoomService roomService,
                              CustomUserDetailsService userDetailsService) {
        this.reservationRepository = reservationRepository;
        this.roomService = roomService;
        this.userDetailsService = userDetailsService;
    }

    public List<ReservationDTO> getReservations(ReservationFilter filter) {
        List<Reservation> reservations;
        if (filter == ReservationFilter.CANCELLED) {
            reservations = reservationRepository.findByStatus(ReservationStatus.CANCELLED);
        } else {
            reservations = reservationRepository.findByStatusNot(ReservationStatus.CANCELLED);
        }
        return reservations.stream().map(this::convertToDTO).toList();

    }

    public List<ReservedDatesDTO> getReservedDatesByRoomId(Long roomId) {
        return reservationRepository.findReservedDatesByRoomId(roomId);
    }

    public ReservationDTO getReservationByIdentifier(String reservationIdentifier) {
        Reservation reservation = reservationRepository.findByReservationIdentifier(reservationIdentifier).orElseThrow(
                () -> new RuntimeException("Reservation doesn't exist")
        );

        return convertToDTO(reservation);
    }

    private String generateReservationCode() {
        return "RSV-" + Long.toString(RANDOM.nextLong() & Long.MAX_VALUE, 36).toUpperCase();
    }

    public ReservationDTO reserveRoom(Long roomId, ReservationRequest request, Authentication authentication) {
        List<ReservedDatesDTO> reservedDates = getReservedDatesByRoomId(roomId);
        for (ReservedDatesDTO reservedRange : reservedDates) {
            if (request.getStartDate().isBefore(reservedRange.getStartDate()) &&
                    request.getEndDate().isAfter(reservedRange.getEndDate())) {
                throw new InvalidReservationRangeException("The selected range includes already reserved dates");
            }
        }
        User user = userDetailsService.loadUserByUsername(authentication.getName());
        Room room = roomService.getRoomById(roomId);
        Reservation reservation =
                Reservation.builder().user(user).startDate(request.getStartDate()).endDate(request.getEndDate())
                        .room(room).reservationIdentifier(generateReservationCode()).build();
        return convertToDTO(reservationRepository.save(reservation));
    }

    public ReservationDTO updateReservation(String reservationIdentifier, ReservationRequest request) {
        Reservation reservationToUpdate = reservationRepository.findByReservationIdentifier(reservationIdentifier)
                .orElseThrow(
                        () -> new RuntimeException("There's no reservation with such reservation identifier")
                );
        reservationToUpdate.setStartDate(request.getStartDate());
        reservationToUpdate.setEndDate(request.getEndDate());
        reservationRepository.save(reservationToUpdate);

        return convertToDTO(reservationToUpdate);
    }

    public void cancelReservation(String reservationIdentifier) {
        Reservation reservation = reservationRepository.findByReservationIdentifier(reservationIdentifier).orElseThrow(
                () -> new RuntimeException("This reservation identifier doesn't correspond to any reservation")
        );

        reservation.setStatus(ReservationStatus.CANCELLED);
        reservationRepository.save(reservation);
    }

    public List<ReservationDTO> getReservationsByUser(Long userId, ReservationFilter filter) {
        List<Reservation> userReservations;
        if (filter == ReservationFilter.CANCELLED) {
            userReservations = reservationRepository.findByUserIdAndStatus(userId, ReservationStatus.CANCELLED);
        } else {
            userReservations = reservationRepository.findByUserIdAndStatusNot(userId, ReservationStatus.CANCELLED);
        }
        return userReservations.stream().map(this::convertToDTO).toList();
    }

    public Reservation getReservationEntityByIdentifier(String reservationIdentifier) {
        return reservationRepository.findByReservationIdentifier(reservationIdentifier).orElseThrow(
                () -> new RuntimeException("This reservation doesn't exist")
        );
    }

    public void updateRequestedModificationState(Long reservationId, ReservationStatus status) {
        Reservation reservation = reservationRepository.findById(reservationId).orElseThrow(
                () -> new ReservationNotFoundException("Reservation with id " + reservationId + " could not be found")
        );

        reservation.setStatus(status);
        reservationRepository.save(reservation);
    }

    private ReservationDTO convertToDTO(Reservation reservation) {
        return new ReservationDTO(
                reservation.getId(),
                reservation.getReservationIdentifier(),
                reservation.getRoom().getId(),
                reservation.getRoom().getName(),
                reservation.getUser().getEmail(),
                reservation.getStartDate(),
                reservation.getEndDate(),
                reservation.getStatus(),
                reservation.getCreatedAt()
        );
    }
}

File: frontend/src/app/services/reservation.service.ts
import { Injectable } from "@angular/core";
import { HttpClient } from "@angular/common/http";
import { Observable } from "rxjs";

export interface ReservationRequest {
    startDate: string;
    endDate: string;
}

export interface Reservation {
        id: number;
        reservationIdentifier: string;
        roomId: number;
        roomName: string;
        userEmail: string;
        startDate: string;
        endDate: string;
        status: string;
        createdAt: Date;
    };

export interface ReservedRange {
    startDate: Date;
    endDate: Date;
}

export type ReservationFilter = 'CANCELLED' | 'NON_CANCELLED';

@Injectable({
    providedIn: 'root'
})
export class ReservationService {

    private readonly apiUrl = 'http://localhost:8080/api/v1';

    constructor(private readonly http: HttpClient){}

    reserveRoom(roomId: number, reservation: ReservationRequest): Observable<Reservation> {
        return this.http.post<Reservation>(this.apiUrl + "/reservations/" + roomId + "/reserve", reservation);
    };

    getReservations(filter: ReservationFilter): Observable<Reservation[]> {
        return this.http.get<Reservation[]>(this.apiUrl + "/reservations", {
                params: filter ? {filter: filter} : {}
            });
    };

    getReservedDates(roomId: number): Observable<ReservedRange[]> {
        return this.http.get<ReservedRange[]>(this.apiUrl + "/reservations/" + roomId + "/reserved-dates");
    };

    getReservationByIdentifier(reservationIdentifier: string): Observable<Reservation> {
        return this.http.get<Reservation>(this.apiUrl + "/reservations/" + reservationIdentifier);
    }

    getReservationsByUserId(userId: string, filter: ReservationFilter): Observable<Reservation[]> {
        return this.http.get<Reservation[]>(this.apiUrl + "/reservations/user/" + userId, {
            params: filter ? {filter: filter} : {}
         });
    }

    updateReservation(reservationIdentifier: string, reservationRequest: Partial<ReservationRequest>): Observable<Reservation> {
        return this.http.patch<Reservation>(this.apiUrl + "/reservations/" + reservationIdentifier, reservationRequest);
    }

    cancelReservation(reservationIdentifier: string): Observable<void> {
        return this.http.delete<void>(this.apiUrl + "/reservations/" + reservationIdentifier);
    }
}
